### ISSUE TITLE
Crash when declaring a datatype const as the result of a function with reads

### ISSUE THREAD

**Issue by danmatichuk** [dafny-lang/dafny#615]
The following program causes Dafny to crash:

```
  class MyClass {
    ghost const repr : object
  }

  datatype D1 = D1(o : MyClass)
  {
    ghost const objs : set<object> := getObjs()

    function getObjs() : set<object>
      reads o
      { {o, o.repr} }
  }
```

The reads clause redundantly includes `o` (since `repr` is marked `const`). However rather than 
giving an error due to having a `reads` clause, Dafny crashes with the following message:

```
Dafny 2.3.0.10506
[ERROR] FATAL UNHANDLED EXCEPTION: System.Diagnostics.Contracts.ContractException: Assertion failed.
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailureImplementation (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0002f] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailure (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x00000] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Diagnostics.Contracts.Contract.ReportFailure (System.Diagnostics.Contracts.ContractFailureKind failureKind, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0003a] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Diagnostics.Contracts.Contract.Assert (System.Boolean condition) [0x00003] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at Microsoft.Boogie.NAryExpr..ctor (Microsoft.Boogie.IToken tok, Microsoft.Boogie.IAppliable fun, System.Collections.Generic.IList`1[T] args, System.Boolean immutable) [0x00043] in <9378f975b0154ba1afae7c7cf97e000c>:0
  at Microsoft.Dafny.Translator+ExpressionTranslator.TrExpr (Microsoft.Dafny.Expression expr) [0x017d1] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator+ExpressionTranslator.TrExpr (Microsoft.Dafny.Expression expr) [0x055f4] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.GetReadonlyField (Microsoft.Dafny.Field f) [0x00452] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.AddClassMembers (Microsoft.Dafny.TopLevelDeclWithMembers c, System.Boolean includeMethods) [0x002c0] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.AddTypeDecl (Microsoft.Dafny.RevealableTypeDecl d) [0x000ca] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.DoTranslation (Microsoft.Dafny.Program p, Microsoft.Dafny.ModuleDefinition forModule) [0x00437] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator+<Translate>d__47.MoveNext () [0x000ea] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.DafnyDriver+<Translate>d__6.MoveNext () [0x00119] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver.Boogie (System.String baseName, System.Collections.Generic.IEnumerable`1[T] boogiePrograms, System.String programId, System.Collections.Generic.Dictionary`2[System.String,Microsoft.Boogie.PipelineStatistics]& statss, Microsoft.Boogie.PipelineOutcome& oc) [0x0012f] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver.ProcessFiles (System.Collections.Generic.IList`1[T] dafnyFiles, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, Microsoft.Dafny.ErrorReporter reporter, System.Boolean lookForSnapshots, System.String programId) [0x00222] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver.ThreadMain (System.String[] args) [0x0003a] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver+<>c__DisplayClass1_0.<Main>b__0 () [0x00001] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00014] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ThreadHelper.ThreadStart () [0x00008] in <f9d1b832704f410aa8ec771f4fe80552>:0
```

**Comment by davidcok** [comment]
Reproduced. test git-issue-615

**Comment by davidcok** [comment]
Fixed by PR#665
